### PR TITLE
Binary ops

### PR DESCRIPTION
**Pull request by gf712**
- [x] Subtract
- [x] Multiply
- [x] Divide
- [x] Logical operators interfaces TODO:
  - [x] add boolean type
  - [x] add binary logical operator -> returns boolean
- [x] AND
- [X] OR
- [X] XOR
- [x] Equals

### CODE REVIEW

File: src/shogun/mathematics/graph/CMakeLists.txt [hunk 1]
```diff
@@ -90,8 +95,13 @@ IF (ENABLE_TESTING)
...
 	ADD_SHOGUN_TEST(nodes/Add_test)
+	ADD_SHOGUN_TEST(nodes/Subtract_test)
+
 	if (UNIX AND NOT APPLE)
 		set_property(TEST Add_test PROPERTY ENVIRONMENT "LD_LIBRARY_PATH=${PROJECT_BINARY_DIR}/src/shogun")
+		set_property(TEST Subtract_test PROPERTY ENVIRONMENT "LD_LIBRARY_PATH=${PROJECT_BINARY_DIR}/src/shogun")
 	endif()
 	add_dependencies(Add_test ${EXCUTORS})
+	add_dependencies(Subtract_test ${EXCUTORS})
+
```
Comment (gf712) [inline src/shogun/mathematics/graph/CMakeLists.txt:139]: @vigsterkr I am guessing this block could become a function/macro?

File: src/shogun/mathematics/graph/CMakeLists.txt [hunk 2]
```diff
@@ -6,6 +6,7 @@ set(SHOGUN_CORE_GRAPH_HEADERS
 	${CMAKE_CURRENT_SOURCE_DIR}/Tensor.h
 	${CMAKE_CURRENT_SOURCE_DIR}/Types.h
 	${CMAKE_CURRENT_SOURCE_DIR}/nodes/Add.h
+	${CMAKE_CURRENT_SOURCE_DIR}/nodes/Subtract.h
```
Comment (gf712) [inline src/shogun/mathematics/graph/CMakeLists.txt:15]: @vigsterkr  I know it is meant to be bad practice in cmake, but maybe we should replace some of these with glob?

Comment (vigsterkr) [reply src/shogun/mathematics/graph/CMakeLists.txt:15]: glob is crazy.... but of course we can do it! we are doing it atm, so it's its all fine

File: src/shogun/mathematics/graph/CMakeLists.txt (hunk 1 above)
Comment (vigsterkr) [reply src/shogun/mathematics/graph/CMakeLists.txt:139]: yep same as all the library setter magic for abi version etc.

File: src/shogun/mathematics/graph/Tensor.h [hunk 3]
```diff
@@ -97,7 +100,8 @@ namespace shogun
 				if (m_shape.size() != shape.size())
 				{
 					error(
-					    "Mismatch in the number of dimensions, expected {}, but "
+					    "Mismatch in the number of dimensions, expected {}, "
+					    "but "
 					    "got {}",
 					    m_shape.size(), shape.size());
```
Comment (gf712) [inline src/shogun/mathematics/graph/Tensor.h:106]: clang-format at it's best...

**Comment by gf712** [comment]
@vigsterkr I think this is good to go. Would rather do the cmake stuff in a separate PR

**Pull request merged**